hadoop-hdds/docs/content/design/truncate.md

```diff
@@ -0,0 +1,146 @@
+
+OM stores `<key, RepeatedTruncateOmKeyInfo>` in truncateTable, and returns succ to client.  The `TruncateKey` in `RepeatedTruncateOmKeyInfo` represents one truncate key operation, so the list of `TruncateKey` allows us to store a list of truncate operations related to one key.
+
+OM starts a background thread to select a certain number of entries from `truncateTable`, and sends these entries to SCM by the `TruncateScmKeyRequestProto`.
+
+### Related New Class
+```
+// include a list of truncate operation
+public RepeatedTruncateOmKeyInfo {
+    private final String volumeName;
+    private final String bucketName;
+    // name of key client specified
+    private String keyName;
+    public List<TruncateKey> truncateKeyList;
+}
+```
+
+### Related New Proto
+```
+/**
+ * A truncate key request sent by OM to SCM, it contains
+ * multiple number of keys (and their blocks needed to be deleted and truncated).
+ */
+message TruncateScmKeyRequestProto {
+     repeated TruncateKey truncateKeys = 1;
+}
+
+/**
+ * A truncated key and its blocks needed to be deleted and the  block needed to be partially truncated.
+ */
+message TruncateKey {
+     required string key = 1;
+     repeated BlockID deleteBlocks = 2;
+     repeated PartialTruncateBlock partialTruncateBlocks = 3;
+}
+
+// A partially truncated block, including the newLength
+message PartialTruncateBlock {
+    required BlockID truncateBlock = 1;
+    required uint64 newLength = 2;
+}
+```
+
+## SCM
+When scm receives `TruncateScmKeyRequestProto`, for the `deleteBlocks`, scm has already implemented the code.
+
+For the `partialTruncateBlocks`, we process it as a transaction like `deleteBlocks`, store `<transactionId, TruncatedBlocksTransaction>` in `truncatedBlocksTable`, and return succ to OM.  
+
+We abstract the code related to delete block transaction, so that truncate and delete blocks can share the abstract code.
+
+SCM starts a background thread to send `TruncateBlocksCommandProto` to the datanode.
+
+### Related New Proto
+```
+// HeartBeat response from SCM, contains a list of block truncate transactions.
+message TruncateBlocksCommandProto {
+  repeated TruncatedBlocksTransaction truncatedBlocksTransactions = 1;
+  required int64 cmdId = 2;
+}
+
+// The truncated blocks which are stored in truncatedBlock.db of scm.
+message TruncatedBlocksTransaction {
+     required int64 txID = 1;
+     required int64 containerID = 2;
+     repeated PartialTruncateBlock blocks = 3;
+     // the retry time of sending truncating command to datanode.
+     required int32 count = 4;
+}
+```
+
+## Datanode
+When Datanode receives `TruncateBlocksCommand`, in the RocksDB, Datanode deletes the chunks which are fully truncated, and updates the chunk length and checksum which is partially truncated. Datanode puts `<TRUNCATING_blockID, newLength>` in RocksDB, and returns succ to SCM.
```

Comment:
In the master branch, container's rocksdb has been divided into 3 different column families.